### ISSUE TITLE
[Bug] SuggestedCacheExpiry can be in the past and results in System.ArgumentOutOfRangeException: The relative expiration value must be positive.

### ISSUE THREAD

**Issue by jennyf19** [AzureAD/microsoft-authentication-library-for-dotnet#2859]
A customer reported this issue with their InMemory OBO cache.

full exception:
```
System.ArgumentOutOfRangeException: The relative expiration value must be positive. (Parameter 'AbsoluteExpirationRelativeToNow')
Actual value was -18869.20:47:50.7160847.
```
Could be that there was an old cache with an expired token, and the expiry computed by MSAL was done in an older version, the newer version of MSAL no longer computes the expiry for OBO tokens, so the cache doesn't know how to process it. not sure, needs to be investigated. Have not been able to repro locally.

**Comment by bgavrilMS** [comment]
Do you know the MSAL version they were using?

**Comment by bgavrilMS** [comment]
What's possibly happening is that CalculateSuggestedCacheExpiry is called in the finally clause. 

So if an exception occured (e.g. during serialization to Redis), we still try to compute the cache expiry, but we will have not added a tokens to the cache. We should check if we indeed added them?

**Comment by jennyf19** [comment]
@petrochuk Can you provide a repro or repro steps? We were never able to reproduce it, or at least I was not, and wasn't sure how we got into that situation to start. 

cc: @bgavrilMS

**Comment by petrochuk** [comment]
I can provide some details. It happens when problematic token gets stuck in local token cache. Customers consistently seeing it until they clear token cache. Here is call stack we are seeing in our telemetry:

```json
[
  [
    {
      "assembly": "System.Core, Version=4.0.0.0",
      "method": "System.Linq.Enumerable.Max"
    },
    {
      "assembly": "Microsoft.Identity.Client, Version=4.39.0.0",
      "method": "Microsoft.Identity.Client.TokenCache.CalculateSuggestedCacheExpiry"
    },
    {
      "assembly": "Microsoft.Identity.Client, Version=4.39.0.0",
      "method": "Microsoft.Identity.Client.TokenCache+<Microsoft-Identity-Client-ITokenCacheInternal-SaveTokenResponseAsync>d__59.MoveNext"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification"
    },
    {
      "assembly": "Microsoft.Identity.Client, Version=4.39.0.0",
      "method": "Microsoft.Identity.Client.Cache.CacheSessionManager+<SaveTokenResponseAsync>d__10.MoveNext"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification"
    },
    {
      "assembly": "Microsoft.Identity.Client, Version=4.39.0.0",
      "method": "Microsoft.Identity.Client.Internal.Requests.RequestBase+<CacheTokenResponseAndCreateAuthenticationResultAsync>d__17.MoveNext"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification"
    },
    {
      "assembly": "Microsoft.Identity.Client, Version=4.39.0.0",
      "method": "Microsoft.Identity.Client.Internal.Requests.InteractiveRequest+<ExecuteAsync>d__8.MoveNext"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification"
    },
    {
      "assembly": "Microsoft.Identity.Client, Version=4.39.0.0",
      "method": "Microsoft.Identity.Client.Internal.Requests.RequestBase+<RunAsync>d__12.MoveNext"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification"
    },
    {
      "assembly": "Microsoft.Identity.Client, Version=4.39.0.0",
      "method": "Microsoft.Identity.Client.ApiConfig.Executors.PublicClientExecutor+<ExecuteAsync>d__2.MoveNext"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw"
    },
    {
      "assembly": "mscorlib, Version=4.0.0.0",
      "method": "System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification"
    },
    {
      "assembly": "bolt.module.auth, Version=1.11.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35",
      "method": "bolt.module.auth.store.AuthTokenStore+<AcquireTokenAsync>d__13.MoveNext"
    }
  ],
  {
    "outerId": "0",
    "message": "Sequence contains no elements",
    "type": "System.InvalidOperationException",
    "id": "46964992"
  }
]
```

**Comment by bgavrilMS** [comment]
Hi @petrochuk  - the "sequence contains no elements" is a different problem. The tracking issue for it is: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3130

I have a PR out of it and we're going to hotfix MSAL, most likely in a day or two. I expect MSAL 4.41.1 to have the fix.

**Comment by bgavrilMS** [comment]
@jennyf19 - I didn't realize that a negative expiration date can crash the distributed cache. I would have expected that that adding an item with a negative TTL to be sematically equivalent to deleting the item. But I guess they decided to throw exception instead. 

Here's what I'm thinking for MSAL fix; 

- if expiration_time < NOW + 5 min, then MSAL will not suggest expiration. We'll still say "HasTokens= false"

**Comment by petrochuk** [comment]
> Hi @petrochuk - the "sequence contains no elements" is a different problem. The tracking issue for it is: #3130
> 
> I have a PR out of it and we're going to hotfix MSAL, most likely in a day or two. I expect MSAL 4.41.1 to have the fix.

Issue #3130 happens on save. We are seeing it coming from CalculateSuggestedCacheExpiry when we try to acquire new token

**Comment by bgavrilMS** [comment]
Will be fixed in next version of MSAL 4.42 or 4.41.1
